synthesizer/src/restrictions/mod.rs

```diff
@@ -0,0 +1,386 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+mod helpers;
+pub use helpers::*;
+
+mod serialize;
+mod string;
+
+use console::{
+    network::prelude::*,
+    program::{Identifier, Literal, Locator, Plaintext, ProgramID},
+    types::Field,
+};
+use ledger_block::{Execution, Input, Output, Transition};
+
+use indexmap::IndexMap;
+
+#[derive(Clone, PartialEq, Eq)]
+pub struct Restrictions<N: Network> {
+    /// The restrictions ID, for the current state of the `Restrictions` list.
+    restrictions_id: Field<N>,
+    /// The set of program IDs that are restricted from being executed.
+    /// e.g. `restricted.aleo` => `..` (all blocks)
+    /// e.g. `restricted.aleo` => `10..` (from block 10 onwards)
+    /// e.g. `restricted.aleo` => `..10` (up to block 10)
+    /// e.g. `restricted.aleo` => `10..20` (from block 10 to block 20)
+    programs: IndexMap<ProgramID<N>, BlockRange>,
+    /// The set of `(program ID, function name)` pairs that are restricted from being executed.
+    /// e.g. `restricted.aleo/foo` => `..` (all blocks)
+    /// e.g. `restricted.aleo/foo` => `10..` (from block 10 onwards)
+    /// e.g. `restricted.aleo/foo` => `..10` (up to block 10)
+    /// e.g. `restricted.aleo/foo` => `10..20` (from block 10 to block 20)
+    functions: IndexMap<Locator<N>, BlockRange>,
+    /// The set of `(program ID, function name, argument)` triples that are restricted from being executed.
+    /// e.g. `restricted.aleo/bar _ aleo1zkpxxxxx _ _` => `..` (all blocks)
+    /// e.g. `restricted.aleo/bar _ aleo1zkpxxxxx _ _` => `10..` (from block 10 onwards)
+    /// e.g. `restricted.aleo/bar _ aleo1zkpxxxxx _ _` => `..10` (up to block 10)
+    /// e.g. `restricted.aleo/bar _ aleo1zkpxxxxx _ _` => `10..20` (from block 10 to block 20)
+    ///
+    /// Note: This design intentionally minimizes the number of total lookups required to check
+    /// for restrictions when a transition matches the described profile. In summary:
+    /// - When a transition does not match the program ID or function name, the total lookup cost is `O(1)`.
+    /// - When a transition matches the program ID & function name, the initial lookup cost is `O(num_inputs + num_outputs)`.
+    ///    - If an input or output index does not match, the additional lookup cost is `0`.
+    ///    - If an input or output index matches, the additional lookup cost is `O(n)` for `n` restricted arguments with the same index.
+    arguments: IndexMap<Locator<N>, IndexMap<ArgumentLocator, IndexMap<Literal<N>, BlockRange>>>,
```

Comment:
How would we handle the case where we want to ban an address from doing anything? It seems structurally more difficult than these checks. I imagine our best option right now would be to restriction `credits.aleo/fee` for some address but this is somewhat easily bypassed as the fee payer can be a different signer.

Reply:
Indeed that would be a nice extension. A first step within the current design could be to extend the check against plaintext literal inputs to also plaintext container inputs. But there's a large design space ahead.